src/WebJobs.Script.Abstractions/Description/IFunctionProvider.cs

```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+using Microsoft.Azure.WebJobs.Script.Description;
+
+namespace Microsoft.Azure.WebJobs.Script.Abstractions
+{
+    public interface IFunctionProvider
```

Comment:
This is basically the same as the existing IFunctionMetadataProvider interface we use internally. For the new public type I think that's the name we want to use. Our existing FunctionMetadataProvider type should then implement this public interface - it becomes a "composite provider" that aggregates all the metadata produced by registered providers. We just need to work out how to load all the custom providers from DI, then replace those with our single composite provider. Perhaps @Fabio has an idea on that.

We can remove the "forceRefresh" caching behavior from this interface, and have it always load fresh, and handle caching upstream where it should be handled, in FunctionMetadataManager which is the primary consumer of this.

Reply:
Or, what if we make `FunctionMetadataProvider` the `DefaultFunctionMetadaProvider` and not a composite provider? The `FunctionMetadaManager` can act as a composite which would take all of the `IFunctionMetadataProvider`s including the `DefaultFunctionMetadaProvider`, any custom ones, and create the collection?
Similar to what I have, but still making it consistent with our internal implementation.

That would mean we remove `IFunctionProvider` and move `IFunctionMetadataProvider` to abstractions and use that for all this, as you suggested.

Reply:
That's a good idea - FunctionMetadataManager can be responsible for aggregating/returning metadata from all registered providers.

Reply:
I discussed this with @fabiocav, and also looked into the design. I think it makes sense to keep the `FunctionMetadataProvider` and custom providers separate.

- One benefit it provides is -- we can keep `bool forcerefresh` only limited to our internal type. We need this force refresh in Provider and not the Manager, as FunctionMetadataManager refreshes its list of providers any time script Host is restarted to account for any changes in such services. By having forcerefresh, we avoid having the provider relook at the filesystem, even though FunctionMetadataManager reiterates over all providers.

- Another key difference is -- `FunctionMetadataProvider` is WebHost scoped. Whereas, these `IFunctionProviders` providers (including proxy) are JobHost scoped and brought it later. They may change over the lifetime of the WebHost.

- I also like the separation of providers with "Custom" or "Virtual" functions, vs actual metadata from filesystem.

To keep such separation, `IFunctionMetadataManager` still acts as a composite, but looks at a single `IFunctionMetadataProvider` and enumerates over all registered `IFunctionProvider` to get the collection of the metadata.

I hope that this is reasonable. We can also discuss this offline too, if you'd like.